File: frontend/supervisor/test/fixtures/merge_path_harness.py
#!/usr/bin/env python3
"""A stand-in for the long-lived Python merge path (ARF-08 test fixture).

The auto-merge backstop is a Python daemon that boots once and stays up. Like
its Node counterpart in ``merge-path-harness.mjs``, this constructs its gate
**once at startup** and then answers a decision per line on stdin, so a test can
flip the gate from outside and watch the answer change in a process it never
restarted.

The pid is in every response for exactly that reason: without it, "the decision
changed" and "the process was replaced" look identical from the outside.
"""

import json
import os
import sys

sys.path.insert(
    0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..", "gate")
)

from arf_gate import ArfGate  # noqa: E402  (path has to be set before the import)

# Constructed once, at boot, exactly like the daemon this stands in for.
GATE = ArfGate(os.environ["ARF_GATE_FILE"])


def main():
    for line in sys.stdin:
        path_id = line.strip()
        if path_id == "":
            continue
        if path_id == "exit":
            break
        decision = GATE.decide_all() if path_id == "all" else GATE.decide(path_id)
        sys.stdout.write(
            json.dumps({"pid": os.getpid(), "runtime": "python", "decision": decision}) + "\n"
        )
        sys.stdout.flush()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: frontend/supervisor/test/fixtures/merge-path-harness.mjs
#!/usr/bin/env node
/**
 * A stand-in for a long-lived Node merge path (ARF-08 test fixture).
 *
 * The watcher runs the `hammer` and `daemon-clean` paths, boots once, and stays
 * up for days. That longevity is the whole reason the config-flag design needs a
 * bounce: the daemon read its governance config at import time and nothing
 * re-reads it.
 *
 * So this harness deliberately does what a real daemon does — constructs its
 * gate **once at startup** and holds it for the life of the process — and then
 * answers a decision per line on stdin. A test can flip the gate from outside
 * and watch the answer change in a process it never restarted, with the pid in
 * every response to prove it.
 */

import { createInterface } from 'node:readline';

import { ArfGate } from '../../../gate/gate-client.mjs';

// Constructed once, at boot, exactly like the daemon this stands in for.
const gate = new ArfGate(process.env.ARF_GATE_FILE);

const lines = createInterface({ input: process.stdin });

lines.on('line', (line) => {
  const pathId = line.trim();
  if (pathId === '') return;
  if (pathId === 'exit') {
    lines.close();
    return;
  }
  const decision = pathId === 'all' ? gate.decideAll() : gate.decide(pathId);
  process.stdout.write(`${JSON.stringify({ pid: process.pid, runtime: 'node', decision })}\n`);
});

lines.on('close', () => process.exit(0));
